types/query/filtered_pagination.go

```diff
@@ -114,3 +119,137 @@ func FilteredPaginate(
 
 	return res, nil
 }
+
+// GenericFilteredPaginate does pagination of all the results in the PrefixStore based on the
+// provided PageRequest. `onResult` should be used to filter or transform the results.
+// `c` is a constructor function that needs to return a new instance of the type T (this is to
+// workaround some generic pitfalls in which we can't instantiate a T struct inside the function).
+// If key is provided, the pagination uses the optimized querying.
+// If offset is used, the pagination uses lazy filtering i.e., searches through all the records.
+// The resulting slice (of type F) can be of a different type than the one being iterated through
+// (type T), so it's possible to do any necessary transformation inside the onResult function.
+func GenericFilteredPaginate[T codec.ProtoMarshaler, F codec.ProtoMarshaler](
+	cdc codec.BinaryCodec,
+	prefixStore types.KVStore,
+	pageRequest *PageRequest,
+	onResult func(key []byte, value T) (F, error),
+	c func() T,
+) ([]F, *PageResponse, error) {
+	// if the PageRequest is nil, use default PageRequest
+	if pageRequest == nil {
+		pageRequest = &PageRequest{}
+	}
+
+	offset := pageRequest.Offset
+	key := pageRequest.Key
+	limit := pageRequest.Limit
+	countTotal := pageRequest.CountTotal
+	reverse := pageRequest.Reverse
+	results := []F{}
+
+	if offset > 0 && key != nil {
+		return results, nil, fmt.Errorf("invalid request, either offset or key is expected, got both")
+	}
+
+	if limit == 0 {
+		limit = DefaultLimit
+
+		// count total results when the limit is zero/not supplied
+		countTotal = true
+	}
+
+	if len(key) != 0 {
+		iterator := getIterator(prefixStore, key, reverse)
+		defer iterator.Close()
+
+		var (
+			numHits uint64
+			nextKey []byte
+		)
+
+		for ; iterator.Valid(); iterator.Next() {
+			if numHits == limit {
+				nextKey = iterator.Key()
+				break
+			}
+
+			if iterator.Error() != nil {
+				return nil, nil, iterator.Error()
+			}
+
+			protoMsg := c()
```

Comment:
Not sure if you tried this already, but from poking around online, seems we could use reflection to perhaps let us generate the underlying struct for the proto type?

```golang
func makeNew[T any](v T) func() any {
    if typ := reflect.TypeOf(v); typ.Kind() == reflect.Ptr {
        elem := typ.Elem()
        return func() any {
            return reflect.New(elem).Interface() // must use reflect
        }
    } else {
        return func() any { return new(T) } // v is not ptr, alloc with new
    }
}
```
(looking at https://stackoverflow.com/questions/71443373/go-function-to-create-either-new-pointer-or-new-value-of-a-generic-type and https://gotipplay.golang.org/p/8mfSV8z7B8p )

Reply:
Yup! @facundomedica and I chatted about this -- I advocated against the use of reflection due to performance reasons. Especially given this is in a hot path (querying). Do you really think reflection is the better approach?

Reply:
My understanding from various posts is that the reflection overhead is really on the order of magnitude of 100ns, which seems to me like it should be acceptable for query hot paths. We'd have to go through around 1000 entries to add a 1ms compute overhead, which shouldn't be the case for these paths. (Plus I think db times are far more dominant)

(Benchmark posts: https://gist.github.com/nfisher/a9aa2d34fedb338667bf , https://stackoverflow.com/questions/46837485/why-is-reflect-type-implements-so-much-slower-than-a-type-assertion , https://github.com/golang/go/issues/7818#issuecomment-396502228 )

I don't have a proper profile of queries, but my gut impression is this is ok and we can make a more performant routine in the future if we need

Reply:
> [Clear is better than clever.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=14m35s)
>
> [Reflection is never clear.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=15m22s)

Reflection in Go is a tool of last resort, to be used only when there is no other option. Whenever you want to reach for `package reflect` it should trigger a little alarm bell in your head :) as it suggests you're doing something a bit... off. I think that's true here.

Concretely, is there a difference between

```go
func GenericFilteredPaginate[T codec.ProtoMarshaler, F codec.ProtoMarshaler](
 	cdc codec.BinaryCodec,
	prefixStore types.KVStore,
	pageRequest *PageRequest,
	onResult func(key []byte, value T) (F, error),
	constructor func() T,
) ([]F, *PageResponse, error) {
```

and

```go
func GenericFilteredPaginate(
 	cdc codec.BinaryCodec,
	prefixStore types.KVStore,
	pageRequest *PageRequest,
	onResult func(key []byte, value codec.ProtoMarshaler) (codec.ProtoMarshaler, error),
	constructor func() codec.ProtoMarshaler,
) ([]codec.ProtoMarshaler, *PageResponse, error) {
```

?

Reply:
Re: Is this a good use of reflection

imo its an example of well encapsulated reflection, the constructor argument is something extremely odd. (Generics should really be amended to make this native in the future)

Re: Is this an improvement, imo yes, the caller no longer has to type cast to and from interfaces, which is necessary for every single query use case. Eliminating unnecessary run-time typecasts / error checks by using generics is a feature for the caller

Reply:
You're trading type assertions for `constructor func() T`, right?

Reply:
yeah

Reply:
```
BEFORE                                      AFTER
========================================    ==========================================
ts = slice of T                             //
Paginate(                                   ts = Paginate(
    ...                                         ...
    func(key, val []byte) error {               func(key []byte, t *T) (*T, error) {                           
        t = parse(val) as type T                    //
        validate t                                  validate t
        append t to ts                              //
        return nil                                  return t, nil
    },                                          },
    ...                                         ...
)                                            )
```

That's definitely different 😅 but I don't think it's obviously better.

Reply:
SGTM